qcodes/dataset/measurements.py

```diff
@@ -159,23 +181,34 @@ def add_result(self,
                                      f'setpoint values for {param}:'
                                      f' {stuffweneed}.'
                                      f' Values only given for {params}.')
+        if inserting_unrolled_array and inserting_as_arrays:
+            raise RuntimeError("Trying to insert multiple data values both "
+                               "in array from and as numeric. This is not "
+                               "possible.")
+        elif inserting_as_arrays:
+            input_size = 1
 
         for index in range(input_size):
             res_dict = {}
             for partial_result in res:
-                param = str(partial_result[0])
-                value = partial_result[1]
-                # For compatibility with the old Loop, setpoints are
-                # tuples of numbers (usually tuple(np.linspace(...))
-                if hasattr(value, '__len__') and not(isinstance(value, str)):
-                    value = cast(Union[Sequence,np.ndarray], value)
-                    if isinstance(value, np.ndarray):
-                        value = np.atleast_1d(value)
-                    res_dict.update({param: value[index]})
-                else:
-                    res_dict.update({param: value})
-
-            self._results.append(res_dict)
+                param_spec = self.parameters[str(partial_result[0])]
+                if param_spec.type == 'array' and index == 0:
+                    res_dict[str(partial_result[0])] = partial_result[1]
```

Comment:
is `res_dict.update(..)` better? (see below)

Reply:
It's significantly slower by my benchmark
```
mydict = {'A': 'Foo', 'B': 'Bar'}
%timeit mydict['C'] = 'FooBar'
35.4 ns ± 1.78 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
%timeit dict['C'] = 'FooBar'
36 ns ± 0.393 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
%timeit mydict.update({'C': 'FooBar'})
159 ns ± 3.36 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Reply:
mmmmmm..... i might not be right, but for the `.update` case, could it be that creation of a new dictionary (`{'C': 'FooBar'}`) takes some time (while in the first two cases, just two strings are created)?

Reply:
It's not all of it 
```
mydict = {'A': 'Foo', 'B': 'Bar'}
myotherdict = {'C': 'FooBar'}
%timeit mydict['C'] = 'FooBar'
%timeit mydict.update({'C': 'FooBar'})
%timeit mydict.update(myotherdict)
34.9 ns ± 0.977 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
154 ns ± 1.9 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
139 ns ± 4.94 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

I don't think update is the right method here. It's really for merging dicts. We only have one element to insert